### PR TITLE
Improve docs for Waker::noop and LocalWaker::noop

### PR DESCRIPTION
**Pull request by ijackson**
* Add a warning about a likely misuse.  (See my commit message for longer rationale.)
 * Apply some probably-accidentally-omitted changes to `LocalWaker`'s docs
 * Add a comment about the clone-and-hack of the docs

I have used [semantic linefeeds](https://rhodesmill.org/brandon/2012/one-sentence-per-line/) for the docs formatting.

### CODE REVIEW

**Comment by tgross35** [comment]
I am not too familiar with these areas of async, so probably best to let somebody else take a look.

r? libs

File: library/core/src/task/wake.rs [hunk 1]
```diff
@@ -531,12 +531,26 @@ impl Waker {
...
     /// This is mostly useful for writing tests that need a [`Context`] to poll
     /// some futures, but are not expecting those futures to wake the waker or
     /// do not need to do anything specific if it happens.
     ///
     /// If an owned `Waker` is needed, `clone()` this one.
     ///
+    /// # Warning
+    ///
+    /// Do not use this function as a way to be able to call `poll` methods,
+    /// in production code,
```
Comment (workingjubilee) [inline library/core/src/task/wake.rs:547]: If we're using semantic linefeeds then let's actually use the 100 characters we're allowed in our doc texts. Also, these commas are redundant.
```suggestion
    /// Do not use this function as a way to be able to call `poll` methods in production code
```

File: library/core/src/task/wake.rs [hunk 2]
```diff
@@ -782,12 +796,28 @@ impl LocalWaker {
         Self { waker }
     }
 
-    /// Creates a new `LocalWaker` that does nothing when `wake` is called.
+    /// Returns a reference to a `LocalWaker` that does nothing when used.
     ///
+    // Note!  Much of the documentation for this method is duplicated
+    // in the docs for `Waker::noop`.
+    // If you edit it, consider editing the other copy too.
+    //
```
Comment (workingjubilee) [inline library/core/src/task/wake.rs:804]: This is pervasively true, I don't feel it's worth a comment? If you want to fix it, then I would prefer a macro that expands into a few doc attrs, so it's actually fixed.

File: library/core/src/task/wake.rs [hunk 3]
```diff
@@ -782,12 +796,28 @@ impl LocalWaker {
...
     ///
+    /// If an owned `LocalWaker` is needed, `clone()` this one.
+    ///
+    /// # Warning
+    ///
+    /// Do not use this function as a way to be able to call `poll` methods,
+    /// in production code,
+    /// from within a non-`async`, non-`poll`, function.
+    ///
+    /// Using a no-op waker for that purpose would cause wakeups to be lost:
```
Comment (workingjubilee) [inline library/core/src/task/wake.rs:817]: "cause wakeups to be lost" feels very imprecise? Who did the wakeups belong to that they are now lost from? "The proper waker", sure, but it feels like it just obscures the meaning of the rest of this sentence.

File: library/core/src/task/wake.rs [hunk 4]
```diff
@@ -541,6 +541,16 @@ impl Waker {
     ///
     /// If an owned `Waker` is needed, `clone()` this one.
     ///
+    /// # Warning
+    ///
+    /// Do not use this function as a way to be able to call `poll` methods,
+    /// in production code,
+    /// from within a non-`async`, non-`poll`, function.
+    ///
+    /// Using a no-op waker for that purpose would cause wakeups to be lost:
```
Comment (kpreid) [inline library/core/src/task/wake.rs:550]: This text is, in my opinion, either incorrect or not specific enough about the situation it is warning against.

`Waker::noop()` is a perfectly good way to _implement executing a task_ in the case where it is an acceptable performance tradeoff, or suitable for the application, to poll the task every time through some sort of loop (or to poll the task exactly once, as in [`now_or_never()`](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.now_or_never)). No wakeups are “lost” because the task is unconditionally awake. This is, to my mind, the main use case of `Waker::noop()`: to implement async execution trivially when it isn't worth the complexity of implementing it better.

It can only cause lost wakeups if you *mix* polling the task this way with polling the task another way.  I get the impression from https://github.com/rust-lang/rust/issues/98286#issuecomment-2243447136 that your use case is perhaps doing something like that, but the text in this change is talking about non-async functions only, not mixing polling from a non-async function and polling from an executor.

Could you rephrase this to be more specific about the situation you are warning against?

**Review by traviscross — CHANGES_REQUESTED** [review]
While I understand the motivation here, I don't think these changes are an improvement.

We should carefully describe what the no-op waker *does*.  We don't need such a strong warning here about what it doesn't do.

Maybe there is a way to say what's trying to be said here, but as it's written, it just feels "off" to me, and I'm not sure it's an important enough point to try to rework this.

File: library/core/src/task/wake.rs (hunk 2 above)
Comment (ijackson) [reply library/core/src/task/wake.rs:803]: I agree that it's pervasively true.  I have seen similar situations in many other places in the rust-lang/rust codebase, often with unintended divergence as we see here.  I'm afraid I don't have the tuits to engage in a campaign of docs deduplication.  I find it difficult to predict the style and taste preferences in the Rust Project.  I feel someone closer to the core of the project should demonstrate how this ought to be done, and probably the libs team should (at least implicitly, by lazy consensus) approve the idiom.  (The docs are not precisely identical, so the macro wouldn't be entirely straightforward.)

IMO the comment is very helpful, even with the remaining duplication, because it tells you *precisely where* you need to edit the clone-and-hack.  If that comment had been in place, perhaps 6f8a944ba4311cbcf5922132721095c226c6fbab wouldn't have contained the slip.  It's  fairly lightweight answer to the problem.

If you don't agree, or want to insist that I do substantially more work instead of this stopgap, I will remove these hunks from this MR branch.

File: library/core/src/task/wake.rs (hunk 4 above)
Comment (ijackson) [reply library/core/src/task/wake.rs:549]: Thanks.  I'll see waht I can do.

I came here to do this because in conversations with colleagues, some of whom are fairly new to Rust, especially async Rust, hadn't understood the implications.

I can see that there can be situations where wakers aren't important, particularly in an *executor*.  I think using a noop waker in a future implementation, without juggling caller(s)' waker(s), is probably a bug.  Am I wrong about that?

Comment (kpreid) [reply library/core/src/task/wake.rs:549]: >  I think using a noop waker in a future implementation, without juggling caller(s)' waker(s), is probably a bug. Am I wrong about that?

Yes, if you are writing a future combinator (that is, `impl Future for...` that delegates to other `Future`s), then you should usually _either_ pass on the provided context's waker, or create your own delegating non-noop wakers (if you want to track which of several child futures woke).

But the text doesn't say “when you are implementing `Future::poll`”. It says “from within a non-`async`, non-`poll`, function”, which *rules out* that case. So, I am confused.

Comment (ijackson) [reply library/core/src/task/wake.rs:550]: > since it involves shared ownership of the task, or using `noop()` within a future combinator

Right.  That is the scenario I'm trying to warn about, particularly using noop within a futures combinator.

I have tried to improve the text to be more specific.  What do you think now?  (I've avoided the term "future combinator" since it seems rather abstruse, and the reader who is about to make the mistake I'm warning about may not realise that that's what they're writing.)

Comment (kpreid) [reply library/core/src/task/wake.rs:550]: > What do you think now?

That’s specific enough that it no longer has the “vague scary warning” problem, yes. Further comments:

* I don't think the words “runtime-agnostic” are adding anything here; if the future has knowledge about the runtime then that potentially overrides *everything* about the conventional `Future`/`Waker` protocol, so it doesn't need to be called out here.

* It’s still possible to use a noop waker correctly in this scenario. In particular, the usual time a wrapper future is likely to be without `Context` is at construction, and futures are always polled once to start. So, if `Foo` wraps `Bar` and polls `Bar` with a noop waker when `Foo` is constructed, then that's fine as long as `Foo` also polls `Bar` with the propagated waker when `Foo` is polled for the first time. I don't have an idea for how to describe this sort of scenario concisely, though.

> I've avoided the term "future combinator" since it seems rather abstruse

Yes, and pedantically, you can have a future that wraps other futures without it being purely a combinator, so this is more correct.

Comment (traviscross) [reply library/core/src/task/wake.rs:550]: I understand what, @ijackson, you're trying to do here, and I appreciate the intent.  And so it kind of saddens me to be so negative on this change, but...

> When implementing a runtime-agnostic future which wraps other future(s), do not use this function as a way to be able to call `poll` methods when you don't have your caller's `Context` available. Using a no-op waker for that purpose would cause wakeups to be lost: the inner future would store only the no-op waker, replacing the proper waker for the waiting task.

In addition to the good points that @kpreid [raised](https://github.com/rust-lang/rust/pull/128064/files#r1700481858), when I read this, I think to myself, "anyone who is able to parse and interpret those two sentences already knows the limitations without needing to be told."

That's what I think the nature of the problem is here.  It seems harder to write an accessible and correct warning than it is to just write an accessible statement of what this does do.

Comment (kpreid) [reply library/core/src/task/wake.rs:550]: Actually, how about something like this? Expanding the current paragraph that says “This is mostly useful for writing tests…”, we could write:

> This is mostly useful for writing tests that need a `Context` to poll some futures, but are not expecting those futures to wake the waker or do not need to do anything specific if it happens. More generally, using `Waker::noop()` to poll a future **discards the notification of when the future should be polled again**, and therefore, it should only be used when such a notification will not be needed to make progress.

I think this would fairly well get across some useful points:

* What you are doing here is discarding a message. (People understand that this can mean that progress stops because the message is not delivered where it is needed.)
* It is reasonable to discard a message specifically when nothing needs to be done about it.

without getting further into the details of precisely classifying what situations this applies to.

Comment (ijackson) [reply library/core/src/task/wake.rs:550]: I like your suggested text.  Thank you!  I will adopt it into the MR.

**Comment by ijackson** [comment]
I've adopted @kpreid's suggestion, with some light edits.  I hope folks like this better.

File: library/core/src/task/wake.rs (hunk 2 above)
Comment (workingjubilee) [reply library/core/src/task/wake.rs:798]: Hmm... I see. I tried to do the refactoring I originally suggested and found it more of a pain in the ass than I expected. As-is, this seems adequate. It makes me vaguely annoyed, which made me initially reluctant about it, but on reflection it is an annoyance with the preexisting state of things.

**Review by workingjubilee — APPROVED** [review]
Yes, I think this is much better as it informs more "positively" than originally framed, answering "What is this _for_?" It's when you don't care about the notification.

I'm good with this assuming @traviscross doesn't have any further comments, though I would appreciate it if you rebased the history to have less of the back-and-forth before we r+ it.

**Review by traviscross — APPROVED** [review]

**Comment by traviscross** [comment]
As @workingjubilee said, this is a better framing.  Thanks for the changes.

File: library/core/src/task/wake.rs [hunk 5]
```diff
@@ -531,10 +531,18 @@ impl Waker {
...
     /// Returns a reference to a `Waker` that does nothing when used.
     ///
+    // Note!  Much of the documentation for this method is duplicated
+    // in the docs for `LocalWaker::noop`.
+    // If you edit it, consider editing the other copy too.
+    //
     /// This is mostly useful for writing tests that need a [`Context`] to poll
     /// some futures, but are not expecting those futures to wake the waker or
     /// do not need to do anything specific if it happens.
     ///
```
Comment (kpreid) [inline library/core/src/task/wake.rs:541]: I think this would read slightly better as a single paragraph, since the "More generally…" is directly referring to what was just introduced, whereas the paragraphs above and below this pair are much more unrelated.

```suggestion
```

File: library/core/src/task/wake.rs [hunk 6]
```diff
@@ -782,12 +790,22 @@ impl LocalWaker {
...
+    /// Returns a reference to a `LocalWaker` that does nothing when used.
     ///
+    // Note!  Much of the documentation for this method is duplicated
+    // in the docs for `Waker::noop`.
+    // If you edit it, consider editing the other copy too.
+    //
     /// This is mostly useful for writing tests that need a [`Context`] to poll
     /// some futures, but are not expecting those futures to wake the waker or
     /// do not need to do anything specific if it happens.
     ///
```
Comment (kpreid) [inline library/core/src/task/wake.rs:802]: (Matching above suggestion.)
```suggestion
```

**Review by kpreid — APPROVED** [review]

File: library/core/src/task/wake.rs (hunk 5 above)
Comment (ijackson) [reply library/core/src/task/wake.rs:541]: I think I prefer it in two paragraphs.  The first paragraph is about tests; then the second paragraph is indeed more general.

I think if we join these together the order should be reversed, so that the general case, with the clear statement about discarded notifications, isn't buried in a paragraph which on first glance seems to be just about testing.

**Comment by ijackson** [comment]
> I would appreciate it if you rebased the history to have less of the back-and-forth before we r+ it.

Right.  Done.

**Comment by workingjubilee** [comment]
Thanks!

@bors r+ rollup
